pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,393 @@
+	"k8s.io/klog/v2"
+
+	invoke "github.com/containernetworking/cni/pkg/invoke"
+	current "github.com/containernetworking/cni/pkg/types/current"
+	netdefclient "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/client/clientset/versioned/typed/k8s.cni.cncf.io/v1"
+	netdefutils "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/utils"
+
+	cniserver "antrea.io/antrea/pkg/agent/cniserver"
+	cnipodcache "antrea.io/antrea/pkg/agent/secondarynetwork/cnipodcache"
+	ipam "antrea.io/antrea/pkg/agent/secondarynetwork/ipam"
+)
+
+const (
+	controllerName = "AntreaAgentPodController"
+	minRetryDelay  = 2 * time.Second
+	maxRetryDelay  = 120 * time.Second
+	numWorkers     = 4
+)
+
+const (
+	networkAttachDefAnnotationKey = "k8s.v1.cni.cncf.io/networks"
+	cniPath                       = "/opt/cni/bin/"
+	sriovInterfaceType            = "sriov"
+	defaultSecondaryInterfaceName = "eth1"
+)
+
+// Set resyncPeriod to 0 to disable resyncing.
+const resyncPeriod = 0 * time.Minute
+
+type PodController struct {
+	kubeClient         clientset.Interface
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface
+	queue              workqueue.RateLimitingInterface
+	podInformer        cache.SharedIndexInformer
+	podLister          corelisters.PodLister
+	nodeName           string
+	podCache           cnipodcache.CNIPodInfoStore
+	cniServer          *cniserver.CNIServer
+}
+
+func NewPodController(kubeClient clientset.Interface,
+	netAttachDefClient netdefclient.K8sCniCncfIoV1Interface,
+	podInformer cache.SharedIndexInformer,
+	resyncPeriod time.Duration,
+	nodeName string,
+	podCache cnipodcache.CNIPodInfoStore,
+	cniServer *cniserver.CNIServer) *PodController {
+
+	pc := PodController{
+		kubeClient:         kubeClient,
+		netAttachDefClient: netAttachDefClient,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "podcontroller"),
+		podInformer:        podInformer,
+		podLister:          corelisters.NewPodLister(podInformer.GetIndexer()),
+		nodeName:           nodeName,
+		podCache:           podCache,
+		cniServer:          cniServer,
+	}
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    pc.enqueuePod,
+			DeleteFunc: pc.enqueuePod,
+			UpdateFunc: pc.enqueuePodUpdate,
+		},
+		resyncPeriod,
+	)
+	return &pc
+}
+
+func podKeyGet(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+func (pc *PodController) enqueuePod(obj interface{}) {
+	var err error
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		podDeletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(err, "Unexpected object received:", obj)
+			return
+		}
+		pod, ok := podDeletedState.Obj.(*corev1.Pod)
+		if !ok {
+			klog.ErrorS(err, "DeletedFinalStateUnknown object is not of type Pod: ", podDeletedState.Obj, pod)
+			return
+		}
+	}
+	podKey := podKeyGet(pod)
```

Comment:
I think we need to ignore all Add/Update/Delete if Status.PodIP is not set, not just Add, and so we probably should check Status.PodIP here, or in the common enqueuePodUpdate() if we still have a common func for all Add/Update/Delete.

Reply:
A correction - as we discussed, we should not ignore Pod Delete event. It can results in "Pod Del (p1)" cleans up "p2"s secondary networks, but let us think about the right solution later.